crypto/evp_extra/evp_asn1.c

```diff
@@ -162,7 +168,27 @@ EVP_PKEY *EVP_parse_private_key(CBS *cbs) {
     return NULL;
   }
 
-  // A PrivateKeyInfo ends with a SET of Attributes which we ignore.
+  // A PrivateKeyInfo & OneAsymmetricKey may optionally contain a SET of Attributes which
+  // we ignore.
+  if (CBS_peek_asn1_tag(&pkcs8, kAttributesTag)) {
+    if (!CBS_get_asn1(cbs, NULL, kAttributesTag)) {
+      OPENSSL_PUT_ERROR(EVP, EVP_R_DECODE_ERROR);
+      return NULL;
+    }
+  }
+
+  int has_pub = 0;
+  // A OneAsymmetricKey may contain an optional PublicKey BIT STRING which is
+  // implicitly encoded. To support public keys that might not be a size
+  // divisible by 8 we leave the first octet of the bit string present, which
```

Comment:
ASN.1 `BIT STRINGS` represent 1's and 0's and of arbitrary length, and as such may not be divisible by eight. So the encoding rules is that the first octet will be a value between 0-7 which represents how many padding bits were added to make the length a multiple of eight. In places we have dealt with bit strings in the past we checked the padding byte immediately and enforce it (usually there is no padding so it should be zero). Assumptions about padding shouldn't be made for all algorithms, so this just calls out that we are leaving padding byte indicator (instead of consuming it), so that a subsequent consumer can validate & use it as appropriate for its algorithm.

Reply:
Just checking: You mean preserving the padding octet in the destination buffer `public_key`?

Reply:
Correct.